### PR TITLE
Simplex support of date objects

### PR DESCRIPTION
**Pull request by cvillamor**

### CODE REVIEW

File: simplex/src/main/java/io/em2m/simplex/std/dates.kt [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    override fun args(args: List<String>) {
+        if (args.isNotEmpty()) {
+            dateFormat = args[0]
+        }
+    }
+
+    override fun transform(value: Any?, context: ExprContext): Any? {
+        val dateInput : Date? = value?.coerce()
+         return if (dateInput != null) {
+             val pattern = DateTimeFormat.forPattern(dateFormat)
```
Comment (christopherlakey) [inline simplex/src/main/java/io/em2m/simplex/std/dates.kt]: The pattern should be parsed in the args function instead of here since that will be cached and therefore more efficient.

File: simplex/src/main/java/io/em2m/simplex/std/dates.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import io.em2m.simplex.model.*
+import io.em2m.simplex.parser.DateMathParser
+import io.em2m.utils.coerce
+import io.em2m.utils.coerceNonNull
+import org.joda.time.format.DateTimeFormat
+import java.util.*
+
+class FormatDatePipe : PipeTransform {
+
+    private var dateFormat: String = ""
```
Comment (christopherlakey) [inline simplex/src/main/java/io/em2m/simplex/std/dates.kt]: This should be a Pattern and have a reasonable default.

File: simplex/src/main/java/io/em2m/simplex/std/dates.kt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+             val pattern = DateTimeFormat.forPattern(dateFormat)
+             pattern.print(dateInput.time)
+         } else value
+    }
+}
+
+class DateMathPipe : PipeTransform {
+
+    private val dateMathParser = DateMathParser()
+    private var dateMath: String = ""
```
Comment (christopherlakey) [inline simplex/src/main/java/io/em2m/simplex/std/dates.kt]: This should be null by default and we should check for null below.  It's actually an error if no value is provided.  We need to decide how to handle errors.

File: simplex/src/main/java/io/em2m/simplex/std/dates.kt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+fun compareDates(n1: Date?, n2: Date?): Int {
+    return when {
+        (n1?.time == n2?.time) -> 0
+        (n2?.time == null) -> 1
+        (n1?.time == null) -> -1
+        else -> n1.compareTo(n2)
+    }
+}
+
+class DateEquals : SingleDateHandler({ k, v -> k == v })
```
Comment (christopherlakey) [inline simplex/src/main/java/io/em2m/simplex/std/dates.kt]: Have you confirmed that date equals works properly?  I'm inclined to suggest we use ```(compareDates(k, v) == 0``` instead of ```k == v```

File: simplex/src/test/java/io/em2m/simplex/ConditionTest.kt [hunk 5]
```diff
@@ -58,4 +56,51 @@ class ConditionTest : Assert() {
         assertTrue(condition.test("true", "true"))
     }
 
+    @Test
+    fun dateEquals() {
+        val condition = DateEquals()
+        val date = Date()
+        assert(condition.test(date, date))
```
Comment (christopherlakey) [inline simplex/src/test/java/io/em2m/simplex/ConditionTest.kt]: Can you add a test here using strings?  Like ```condition.test("2015-04-20", "2015-04-20")```?  We should also test for a negative.

File: simplex/src/main/java/io/em2m/simplex/std/dates.kt (hunk 4 above)
Comment (cvillamor) [reply simplex/src/main/java/io/em2m/simplex/std/dates.kt]: This should be ok. we take the time out of both dates when we do the comparison. Ive added additional unit tests around this function.
